### ISSUE TITLE
Forbid to name your variables `self`, `cls`, and `mcs`

### ISSUE THREAD

**Issue by sobolevn** [wemake-services/wemake-python-styleguide#298]
# Rule request

## Thesis

Currently you can define variables named `self`, `cls`, and `mcs` with no error:

```python

self = 12
cls = 89
mcs = 1

class Test(object):
    def __init__(self):
        self = 16
```

## Reasoning

These names are magical. All of them have the perfectly defined context and use-case.
These names should be reserved for method parameters only.

**Comment by AlexArcPy** [comment]
@sobolevn I've just added them to the `\wemake_python_styleguide\constants.py` in the `VARIABLE_NAMES_BLACKLIST` https://github.com/wemake-services/wemake-python-styleguide/blob/master/wemake_python_styleguide/constants.py#L59

```
    # Used only within a certain context:
    'cls',
    'mcs',
    'self',
```

This should be it? Or am I missing anything?

**Comment by AlexArcPy** [comment]
Hm it seems as it yells at the line `def __init__(self):` in

```python
class Test(object):
    def __init__(self):
        self = 16
```

`code_for_flake8_to_analyze.py:7:5: Z110 Found wrong variable name "self"`

Any ideas how we could workaround this still using the blacklisted variables constants?

**Comment by AlexArcPy** [comment]
@sobolevn having a function with an argument named any of those, should this be recorded as a violation?

```python
def my_func(cls):
    print(cls)
```

What about instance (with `self`) and class (with `cls`) methods?

```python
class MyClass(object):
    def __init__(self, value1):
        self.value1 = 10
    def method1(self, mcs):
        print(mcs * 2)
```

Should having `mcs` in the `def method1(self, mcs):` raise a violation?

Then I think it may be necessary to modify the https://github.com/wemake-services/wemake-python-styleguide/blob/bfd3a13efe7534e155dd325f419d884c4c667a15/wemake_python_styleguide/visitors/ast/naming.py#L68 `_check_function_signature` to somehow understand whether this is an instance/class method or just a stand-alone function and have different checks for `self` in particular depending on whether it is a class/instance method or not.

**Comment by sobolevn** [comment]
I think that this should be a separate rule. Not regular `WrongVariableName`. Why?

1. It will have different (and fixed) set of values that are covered. Only: `self`, `cls`, and `mcs`
2. It will work based on different logic: it will only allow these three names to be the first function argument, otherwise raising an error

So, that's how I see it. Does it sound reasonable to you?
